wormhole-connect/src/hooks/useRoutesQuotesBulk.ts

```diff
@@ -0,0 +1,82 @@
+import { useState, useEffect, useMemo } from 'react';
+import { Chain, routes } from '@wormhole-foundation/sdk';
+
+import config from 'config';
+
+type RoutesQuotesBulkParams = {
+  sourceChain?: Chain;
+  sourceToken: string;
+  destChain?: Chain;
+  destToken: string;
+  amount: string;
+  nativeGas: number;
+};
+
+type QuoteResult = routes.QuoteResult<routes.Options>;
+
+type HookReturn = {
+  quotesMap: Record<string, QuoteResult | undefined>;
+  isFetching: boolean;
+};
+
+const useRoutesQuotesBulk = (
+  routes: string[],
+  params: RoutesQuotesBulkParams,
+): HookReturn => {
+  const [isFetching, setIsFetching] = useState(false);
+  const [quotes, setQuotes] = useState<QuoteResult[]>([]);
+
+  useEffect(() => {
+    let unmounted = false;
+    if (
+      !params.sourceChain ||
+      !params.sourceToken ||
+      !params.destChain ||
+      !params.destToken ||
+      !params.amount
+    ) {
+      return;
+    }
+
+    // Forcing TS to infer that fields are non-optional
+    const rParams = params as Required<RoutesQuotesBulkParams>;
+
+    setIsFetching(true);
+    config.routes
+      .computeMultipleQuotes(routes, rParams)
+      .then((quoteResults) => {
+        if (!unmounted) {
+          setQuotes(quoteResults);
+          setIsFetching(false);
+        }
+      });
+
+    return () => {
+      unmounted = true;
+    };
+  }, [
+    routes.join(),
+    params.sourceChain,
+    params.sourceToken,
+    params.destChain,
+    params.destToken,
+    params.amount,
+    params.nativeGas,
+  ]);
+
+  const quotesMap = useMemo(
+    () =>
+      routes.reduce((acc, route, index) => {
+        acc[route] = quotes[index];
+        return acc;
+      }, {} as Record<string, QuoteResult | undefined>),
+    [routes.join(), quotes],
```

Comment:
What's the difference to use `routes.join()` vs. `routes` itself here?

Reply:
`routes` is an array, while joining it's becoming a string, essentially `useMemo` becomes agnostic of `routes` reference changes, and only re-calculates when actual value is changed, helps preserve `quote` references in the components that rely on them. It's a linear time operation, but implementation is native in browsers and is incredibly fast. (Actual string concatenation is executed in native runtime, not in JS VM)

Reply:
I see, so when it's an array/object dependencies are compared by reference:

> React will compare each dependency with its previous value using the [Object.is](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) comparison.